core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala

```diff
@@ -221,5 +222,15 @@ private class CompactedEventLogFileWriter(
     hadoopConf: Configuration)
   extends SingleEventLogFileWriter(appId, appAttemptId, logBaseDir, sparkConf, hadoopConf) {
 
-  override val logPath: String = originalFilePath.toUri.toString + EventLogFileWriter.COMPACTED
+  override val shouldCompress = EventLogFileWriter.codecName(originalFilePath).isDefined
+
+  override val compressionCodec =
+    if (shouldCompress) {
+      val originalCodecName = EventLogFileWriter.codecName(originalFilePath).getOrElse("")
+      Some(CompressionCodec.createCodec(sparkConf, originalCodecName))
+    } else {
+      None
+    }
+
+  override def logPath: String = originalFilePath.toUri.toString + EventLogFileWriter.COMPACTED
```

Comment:
Since this is not changed, please put this before all new additions. I guess that will reduce your PR's diff.

Reply:
It seems that we don't need to change. Shall we revert the change on this line? WDYT?

Reply:
> ~Since this is not changed, please put this before all new additions. I guess that will reduce your PR's diff.~ Is this changed from `val` to `def`?

yes, here change to def to avoid NPE. in addition, i changed `compressionCodecName` to lazy val in class `EventLogFileWriter`.

Reply:
> ~Since this is not changed, please put this before all new additions. I guess that will reduce your PR's diff.~ Is this changed from `val` to `def`?

Here is an example,  it reproduce and explain why is my abstract or overridden val null. 
https://docs.scala-lang.org/tutorials/FAQ/initialization-order.html